Compare and Equal

NathanReb opened the pull request:
Fixes #7 

This PR adds `compare` and `equal` functions.

Implementation is the same as stdlib's `Set.compare`. Due to the stronger invariants in DIET, I think it might be possible to optimize it further by traversing both trees and stopping if two incompatible intervals (ie which intersect or whose bounds are adjacent) are met but I felt like this was a good enough first version.

I added a few tests as well as a fuzz check and some benchmarks.

Here are the benchmarks results with a naive pair of folds and a `List.equal`:
```
$ dune exec -- bench/bench.exe -quota 5s
Estimated testing time 30s (6 benchmarks x 5s). Change using '-quota'.
┌───────────────────────┬─────────────┬────────────┬──────────┬──────────┬────────────┐
│ Name                  │    Time/Run │    mWd/Run │ mjWd/Run │ Prom/Run │ Percentage │
├───────────────────────┼─────────────┼────────────┼──────────┼──────────┼────────────┤
│ Equal (size 10)       │    290.16ns │    120.00w │          │          │      0.80% │
│ Equal (size 100)      │  2_843.26ns │  1_200.03w │    2.75w │    2.75w │      7.79% │
│ Equal (size 1000)     │ 36_488.74ns │ 12_000.28w │  279.07w │  279.07w │    100.00% │
│ Not equal (size 10)   │    169.39ns │    120.00w │          │          │      0.46% │
│ Not equal (size 100)  │  1_635.65ns │  1_200.03w │    2.74w │    2.74w │      4.48% │
│ Not equal (size 1000) │ 22_713.03ns │ 12_000.28w │  278.92w │  278.92w │     62.25% │
└───────────────────────┴─────────────┴────────────┴──────────┴──────────┴────────────┘
```
and here the results with the actual `equal`:
```
$ dune exec -- bench/bench.exe -quota 5s
Estimated testing time 30s (6 benchmarks x 5s). Change using '-quota'.
┌───────────────────────┬─────────────┬────────────┬──────────┬──────────┬────────────┐
│ Name                  │    Time/Run │    mWd/Run │ mjWd/Run │ Prom/Run │ Percentage │
├───────────────────────┼─────────────┼────────────┼──────────┼──────────┼────────────┤
│ Equal (size 10)       │    245.14ns │    140.01w │          │          │      0.76% │
│ Equal (size 100)      │  2_471.80ns │  1_400.07w │    0.23w │    0.23w │      7.70% │
│ Equal (size 1000)     │ 32_084.88ns │ 14_000.75w │    3.37w │    3.37w │    100.00% │
│ Not equal (size 10)   │     92.12ns │     56.00w │          │          │      0.29% │
│ Not equal (size 100)  │     86.83ns │     98.01w │          │          │      0.27% │
│ Not equal (size 1000) │    113.15ns │    140.01w │          │          │      0.35% │
└───────────────────────┴─────────────┴────────────┴──────────┴──────────┴────────────┘
```

Let me know what you think!

NathanReb committed:
Add equal and compare functions

NathanReb committed:
Add benchmarks for equal

emillon commented on bench/bench.ml:
```diff
@@ -0,0 +1,59 @@
+    let to_string = string_of_int
+  end)
+
+let state = Random.State.make [|994326685; 290180182; 366831641|]
```
Where do these constants come from? Would `self_init` be enough here?

NathanReb replied:
I just generated them using `Random.int (2^30 -1)`. The whole point is to make sure the whole benchmarks suite is deterministic and reproducible, which it won't be with `self_init`.

A comment could maybe clarify that.

emillon replied:
OTOH, the benchmark will run a loop, so if you use random numbers (rather than fixed ones) you'll get a better idea of the mean performance (compared to the performance on this particular input).

NathanReb replied:
I think it's not incompatible. We can generate a series of random diet but with the same PRNG state so that you have a good idea of the average perfs while still ensuring it makes sense to compare different runs of the benchmarks.

What do you think?

emillon replied:
If there are enough runs I don't think there's a difference, and as `core_bench` docs use `self_init` I'd use that, but as you prefer, that's not the important part of this PR.

NathanReb replied:
Ok, that sounds good to me.

It is slightly annoying that we can't use a combination of `create_indexed` and `create_with_initialization` though as my understanding is that in the first case we'll just repeatedly run equal on a randomly generated pair of diets where in the second we can generate a new pair of diets for each run.

I suggest we use `create_with_initialization` otherwise it could really raise inconsistent results across different runs.

NathanReb replied:
I pushed something. Let me know what you think!

emillon commented on bench/bench.ml:
```diff
@@ -0,0 +1,59 @@
+
+let state = Random.State.make [|994326685; 290180182; 366831641|]
+
+let shuffle a =
```
How about renaming it to `fisher_yates_shuffle` to avoid readers check that it's a correct shuffle? :slightly_smiling_face:

NathanReb replied:
sure!

emillon commented on bench/bench.ml:
```diff
@@ -0,0 +1,59 @@
+    (Bench.make_command
+       [ Bench.Test.create ~name:"Equal (size 10)" (fun () -> ignore @@ IntDiet.equal d10 d10')
+       ; Bench.Test.create ~name:"Equal (size 100)" (fun () -> ignore @@ IntDiet.equal d100 d100')
+       ; Bench.Test.create ~name:"Equal (size 1000)" (fun () -> ignore @@ IntDiet.equal d1000 d1000')
```
`core_bench` seems to support that a bit better with [`create_indexed`](https://github.com/janestreet/core_bench/blob/49024d40e7d098eedf43f5c6437df3070c646204/src/bench.mli#L62), that might be worth it.

NathanReb replied:
ok, I'll look into it!

NathanReb committed:
Add a comment to justify PRNG state initialization

NathanReb committed:
Rename shuffle function to a more explicit fisher_yates_shuffle

NathanReb committed:
Generate new diet pairs for each run

NathanReb committed:
Use self initialized state instead of a fixed one

emillon commented on bench/bench.ml:
```diff
@@ -0,0 +1,73 @@
+  end)
+
+(** This fix state ensures that the benchmarks are reproducible.
+    Those constants have been generated using [Random.int 1_073_741_823].
```
The constants are not there anymore so we can remove this line

```suggestion
```

Julow commented:
Nice !

This can be a little faster by scarifying the elegant type `enum` and using a list of `node` instead. ([code](https://github.com/NathanReb/ocaml-diet/compare/equality-function...Julow:16_using_list))

```
Before:
┌─────────────────────┬─────────────┬────────────┬──────────┬──────────┬────────────┐
│ Name                │    Time/Run │    mWd/Run │ mjWd/Run │ Prom/Run │ Percentage │
├─────────────────────┼─────────────┼────────────┼──────────┼──────────┼────────────┤
│ Equal/size 10       │    150.96ns │    140.01w │          │          │      0.62% │
│ Equal/size 100      │  1_509.06ns │  1_400.07w │    0.22w │    0.22w │      6.16% │
│ Equal/size 1000     │ 24_513.00ns │ 14_000.75w │    3.45w │    3.45w │    100.00% │
│ Not equal/size 10   │     28.54ns │     42.00w │          │          │      0.12% │
│ Not equal/size 100  │     55.85ns │     98.01w │          │          │      0.23% │
│ Not equal/size 1000 │     77.31ns │    140.01w │          │          │      0.32% │
└─────────────────────┴─────────────┴────────────┴──────────┴──────────┴────────────┘
After:
┌─────────────────────┬─────────────┬───────────┬──────────┬──────────┬────────────┐
│ Name                │    Time/Run │   mWd/Run │ mjWd/Run │ Prom/Run │ Percentage │
├─────────────────────┼─────────────┼───────────┼──────────┼──────────┼────────────┤
│ Equal/size 10       │    135.65ns │    60.00w │          │          │      0.59% │
│ Equal/size 100      │  1_369.21ns │   600.01w │          │          │      5.92% │
│ Equal/size 1000     │ 23_141.91ns │ 6_000.14w │    0.64w │    0.64w │    100.00% │
│ Not equal/size 10   │     24.74ns │    18.00w │          │          │      0.11% │
│ Not equal/size 100  │     48.73ns │    42.00w │          │          │      0.21% │
│ Not equal/size 1000 │     45.17ns │    60.00w │          │          │      0.20% │
└─────────────────────┴─────────────┴───────────┴──────────┴──────────┴────────────┘
```

NathanReb committed:
Optimize compare by using a node list instead of the enum type

NathanReb commented:
Nice optimization!

I rebased, rewrote the tests a bit differently so it takes advantage of OUnit's nesting and rewrote enum/compare using lists of nodes!

emillon commented:
Let's get this merged, thanks!
